### PR TITLE
Add support for bindings to `RabbitMQAPI`

### LINKED ISSUE #145: Configuration script needs to add bindings to support delayed messages exchanges
`RabbitMQAPI` in `zocalo.util.rabbitmq` needs extending to support bindings to allow the configuration of queues bound to exchanges other than the default exchange

### PR DESCRIPTION
**Pull request by d-j-hatton**
Bindings are handled slightly differently to queues and exchanges in the RabbitMQ HTTP API. A `POST` is required rather than a `PUT` to the address `api/bindings/e/{exchange name}/{e, q}/{queue or exchange name}`. The `routing_key` and `arguments` can be declared in the body and a "name" (called the `properties_key`) is then created for the binding from the `routing_key` and a hash of `arguments`. A single binding can then be found at `api/bindings/e/{exchange name}/{e, q}/{queue or exchange name}/{properties_key}`. The proposed API allows for deletion of a single binding if `properties_key` is specified or of all bindings between a given source and destination otherwise.

Resolves #145

### CODE REVIEW

File: src/zocalo/util/rabbitmq.py [hunk 1]
```diff
@@ -624,11 +655,73 @@ def put(
...
+        vhost: Optional[str] = None,
+        source: Optional[str] = None,
+        destination: Optional[str] = None,
+        destination_type: Optional[DestinationType] = None,
+    ) -> List[BindingInfo]:
+        endpoint = "bindings"
+        if vhost is not None:
+            endpoint = f"{endpoint}/{vhost}"
+        _check = (source, destination, destination_type)
+        if not all(_ is None for _ in _check) and None in _check:
```
Comment (Anthchirp) [inline src/zocalo/util/rabbitmq.py:684]:
```suggestion
        _check = {source, destination, destination_type}
        if None in _check and len(_check) > 1:
```

**Review by Anthchirp — APPROVED** [review]
I think I would prefer if we changed our use of enums to be more pythonic, ie. have meaningful names in capitals, and the RabbitMQ-specific stringy bits as values.

ie.
```python
class DestinationType(enum.Enum):
    QUEUE = "q"
    EXCHANGE = "e"
```
and then use eg. `DestinationType.QUEUE` to get an enum value and `DestinationType.QUEUE.value` to get the rabbitystring

File: src/zocalo/util/rabbitmq.py [hunk 2]
```diff
@@ -12,7 +12,7 @@
 
 import zocalo.configuration
 
-logger = logging.getLogger("workflows.transport.pika_transport")
+logger = logging.getLogger("zocalo.util.rabbitmq")
```
Comment (rjgildea) [inline src/zocalo/util/rabbitmq.py:15]: Good spot :+1:

File: src/zocalo/util/rabbitmq.py [hunk 3]
```diff
@@ -624,11 +655,73 @@ def put(
...
+        endpoint = f"bindings/{vhost}/e/{source}/{destination_type.name}/{destination}"
+        if properties_key is None:
+            props = [BindingInfo(**r).properties_key for r in self.get(endpoint).json()]
+        else:
+            props = [properties_key]
+        for prop in props:
+            resp = self.delete(f"{endpoint}/{prop}")
+            if resp.status_code == 404:
+                logger.error(f"404 not found when deleting {endpoint}/{prop}")
+            if resp.status_code == 405:
```
Comment (rjgildea) [inline src/zocalo/util/rabbitmq.py:722]:
```suggestion
            elif resp.status_code == 405:
```

**Review by rjgildea — APPROVED** [review]

**Pull request merged**